[quant] Add default_qat_config_v2 to use the fused observer+fake_quant module

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62702
* #62863

Summary:
Expose the qconfig to the user to speed up training by leveraging the fused module.
The module currently supports per-tensor/per-channel moving avg observer and fake-quantize.

For details on perf benefits, refer to https://github.com/pytorch/pytorch/pull/61691

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093719](https://our.internmc.facebook.com/intern/diff/D30093719)

supriyar committed:
[quant] Add default_fused_qat_config to use the fused observer+fake_quant module
Summary:
Expose the qconfig to the user to speed up training by leveraging the fused module.
The module currently supports per-tensor/per-channel moving avg observer and fake-quantize.

Note: reduce_range can be specified by modifying the quant_min/quant_max argument in the args.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093719).

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -343,6 +343,19 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
                                                       qscheme=torch.per_tensor_affine,
                                                       reduce_range=True)
 
+default_fused_act_fake_quant = FusedMovingAvgObsFakeQuantize.with_args(observer=MovingAverageMinMaxObserver,
```
I'd vote for using `reduce_range` to be consistent with other qconfigs, and change it only when we change the other qconfigs, to keep everything consistent.  There is not clear timeline to deprecate it, and it would be confusing to users if some qconfigs used it and some did not.

supriyar replied:
added reduce_range in #62863

vkuzo commented on torch/quantization/qconfig.py:
```diff
@@ -93,6 +93,8 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
 default_activation_only_qconfig = QConfig(activation=default_fake_quant,
                                           weight=torch.nn.Identity)
 
+default_fused_qat_config = QConfig(activation=default_fused_act_fake_quant, weight=default_fused_wt_fake_quant)
```
the fact that this is fused sounds like an implementation detail that the user shouldn't need to know.  Can we instead name it something like `default_qat_qconfig_v2`, and document the difference in the comments?  Or, should we figure out how to version default qconfigs first and then land this PR afterwards?

supriyar replied:
makes sense, for now I've named it `default_qat_qconfig_v2` while we think about versioning support for qconfigs.

supriyar committed:
Update on "[quant] Add default_fused_qat_config to use the fused observer+fake_quant module"


Summary:
Expose the qconfig to the user to speed up training by leveraging the fused module.
The module currently supports per-tensor/per-channel moving avg observer and fake-quantize.

Note: reduce_range can be specified by modifying the quant_min/quant_max argument in the args.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093719](https://our.internmc.facebook.com/intern/diff/D30093719)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093719).

raghuramank100 commented on torch/quantization/qconfig.py:
```diff
@@ -93,6 +93,10 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
 
+# QAT config that uses a fused observer + fake quant modules for optimized training performance.
+# to modify the activation/weight observers, the default entries in fake_quantize.py can be modified.
+default_qat_qconfig_v2 = QConfig(activation=default_fused_act_fake_quant, weight=default_fused_wt_fake_quant)
```
Lets have a new version in the get_default_qat_qconfig function, this way we can support both fbgemm and qnnpack and also roll this out easily by bumping the version number
``` python
# Move version to 1 when we want to roll this out.
def get_default_qat_qconfig(backend='fbgemm', version=None):
    # Histogram observer is too slow for quantization aware training

    if backend == 'fbgemm':
      if version is None:
          qconfig = QConfig(activation=FakeQuantize.with_args(observer=MovingAverageMinMaxObserver,
                                                            quant_min=0,
                                                            quant_max=255,
                                                            reduce_range=True),
                          weight=default_per_channel_weight_fake_quant)
      elif backend == 'qnnpack':
          qconfig = QConfig(activation=FakeQuantize.with_args(observer=MovingAverageMinMaxObserver,
                                                            quant_min=0,
                                                            quant_max=255,
                                                            reduce_range=False),
                          weight=default_weight_fake_quant)
 if version ==1:
  if backend =='fbgemm':
      qconfig = QConfig(activation=default_fused_act_fake_quant, weight=default_fused_wt_fake_quant)
if backend=='qnnpack':
    ....
    else:
        qconfig = default_qat_qconfig
    return qconfig
```

raghuramank100 reviewed (changes requested):
Please see comments

supriyar committed:
Update on "[quant] Add default_qat_config_v2 to use the fused observer+fake_quant module"


Summary:
Expose the qconfig to the user to speed up training by leveraging the fused module.
The module currently supports per-tensor/per-channel moving avg observer and fake-quantize.

Note: reduce_range can be specified by modifying the quant_min/quant_max argument in the args.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093719](https://our.internmc.facebook.com/intern/diff/D30093719)

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Will be good to announce this change in the OSS release notes and internally

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093719).

supriyar committed:
Update on "[quant] Add default_qat_config_v2 to use the fused observer+fake_quant module"


Summary:
Expose the qconfig to the user to speed up training by leveraging the fused module.
The module currently supports per-tensor/per-channel moving avg observer and fake-quantize.

Note: reduce_range can be specified by modifying the quant_min/quant_max argument in the args.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093719](https://our.internmc.facebook.com/intern/diff/D30093719)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093719).
